### PR TITLE
[2.14] all.sh: make it possible to run a subset of the components

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Break up `all.sh` into components and make it possible to run a subset of the components by listing them on the command line.

To this effect:

1. Reorganize `all.sh` into functions.
2. A very minimal amount of tidying up. I did remove one redundant component.
3. All components back up `config.h` systematically.
4. If you pass component names on the command line, `all.sh` only executes these components. With `--except`, `all.sh` runs all the components except the ones named on the command line. With no non-option argument, `all.sh` executes all the components. Run `all.sh --list-components` to list available components.

Apart from some minor differences in the lead-in and clean-up to each component and some minor component reorganization, the result and reporting of `all.sh` should be the same as before.

Potential follow-ups, which are explicitly out of scope in this PR:

* Simplify keep-going mode: if a component fails in the middle, just move on to the next component.
* Make the reporting fit the component model better: report the names of failing components, report the number of passing components...
* Self-checks to ensure that when someone adds a component, they don't forget to add it to the default list.
* Aids to parallelization. The primary goal of this PR is to pave the way for parallelization, not to achieve parallelization.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Build that should succeed: https://jenkins-internal.mbed.com/job/mbedtls-release/832/

Build that should report failures: https://jenkins-internal.mbed.com/job/mbedtls-release/833/

**Comment by gilles-peskine-arm** [comment]
I just amended the commit that added `--except` because it broke non-except usage (`all.sh` with an explicit list of components).

**Comment by gilles-peskine-arm** [comment]
New runs of `all.sh` based on 15992856d87d648d89aeaf89b673355bcd999021: https://jenkins-internal.mbed.com/job/mbedtls-release/850/ (should pass), https://jenkins-internal.mbed.com/job/mbedtls-release/851/ (should fail `RSA_NO_CRT` and `check-files`).

File: tests/scripts/all.sh [hunk 1]
```diff
@@ -55,6 +55,16 @@
 # Notes for maintainers
 # ---------------------
 #
+# The bulk of the code is organized into functions that follow one of the
+# following naming conventions:
+#  * pre_XXX: things to do before running the tests, in order.
+#  * run_XXX: independent components. They can be run in any order.
```
Comment (mpg) [inline tests/scripts/all.sh]: I don't think `run_XXX` is correct here. You probably mean `component_TYPE_XXX`.

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -483,16 +483,15 @@ component_test_default_cmake_gcc_asan () {
...
-
-    msg "build: with ASan (rebuild after ref-configs)" # ~ 1 min
-    make
-
     msg "test: compat.sh (ASan build)" # ~ 6 min
     if_build_succeeded tests/compat.sh
 }
 
+component_test_ref_configs () {
+    msg "test/build: ref-configs (ASan build)" # ~ 6 min 20s
```
Comment (mpg) [inline tests/scripts/all.sh]: The `(ASan build)` part of the message is no longer correct. Please either remove it from the message, or add the following line before running the script:
```
CC=gcc cmake -D CMAKE_BUILD_TYPE:String=Asan .
```

File: tests/scripts/all.sh [hunk 3]
```diff
@@ -145,6 +159,11 @@ General options:
   -k|--keep-going       Run all tests and report errors at the end.
   -m|--memory           Additional optional memory tests.
      --armcc            Run ARM Compiler builds (on by default).
+     --except           If some components are passed on the command line,
+                        run all the tests except for these components. In
+                        this mode, you can pass shell wildcard patterns tha
```
Comment (mpg) [inline tests/scripts/all.sh]: Typo: `tha`

File: tests/scripts/all.sh [hunk 4]
```diff
@@ -145,6 +159,11 @@ General options:
...
+                        component names, e.g. "$0 --except 'test_*'" to
+                        exclude all components that run tests.
```
Comment (mpg) [inline tests/scripts/all.sh]: The commit contents and the commit message seem to give a different view of this option. From the commit content it looks like the second `--except` in the commit message is redundant, as usage would be `--except --potentially-other-options-here foo bar baz`.

File: tests/scripts/all.sh (hunk 1 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: Fixed by amending "Move the code into functions. No behavior change."

File: tests/scripts/all.sh (hunk 2 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: Fixed by amending "Move test-ref-configs into its own component"

File: tests/scripts/all.sh (hunk 3 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: Fixed by amending "Add --except mode: run all components except a list"

File: tests/scripts/all.sh (hunk 4 above)
Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: Fixed by rewording the commit message

**Comment by gilles-peskine-arm** [comment]
@mpg Since one of the issues you pointed out was in a commit message, and all of them were minor errors with no lasting interest, I rewrote the history to fix each issue in the commit that introduced it. I did not change the base of the branch, so it should be easy to review the differences. The previous history is in https://github.com/gilles-peskine-arm/mbedtls/tree/all_sh-2.14-2

Because the first commit is tedious and error-prone, I will not redo it on top of the latest branch. Instead I will merge `development` into this branch and do the changes made in `development` since this branch was spun off.

**Comment by gilles-peskine-arm** [comment]
I decided to make `config.h` restoration more systematic in this PR after all. I had initially decided not to, because I wanted to minimize the changes. But [having it break was annoying](https://github.com/ARMmbed/mbedtls/pull/1643/commits/ee427b26d4ec99a67569637cad833a1ab5247094), and it turns out to be a simple change. I was afraid of special cases with auxiliary scripts that do their own backup, but that's ok, all of them cope if `config.h.bak` already exists.

**Comment by gilles-peskine-arm** [comment]
Release job (should pass): https://jenkins-internal.mbed.com/job/mbedtls-release/907/

**Comment by gilles-peskine-arm** [comment]
This PR targets 2.14, which at the time I made it was supposed to become an LTS branch, but the LTS was delayed until 2.16. I'll reuse the history of this PR to make a PR for 2.16 (with a merge of `mbedtls-2.16.0`) and for mbedcrypto (which forked at 2.14).

**Comment by gilles-peskine-arm** [comment]
I forgot a reason not to do the systematic back-up of `config.h`: it adds some differences in the log. I tried to keep these to a minimum. But given that since I made this PR we did goof with this backup, I now think that the risk of forgetting a backup in a component outweighs the risk that we'll miss some important change in the test logs.

**Comment by gilles-peskine-arm** [comment]
The 2.16 and development version of this PR is up at #2325

**Comment by mpg** [comment]
I checked the result of the CI run, it only failed in an expected place (GnuTLS fragmentation).

**Comment by sbutcher-arm** [comment]
@gilles-peskine-arm, why is this branch still open if it has been superseded by PR #2325?

**Comment by gilles-peskine-arm** [comment]
Since https://github.com/ARMmbed/mbedtls/pull/2325 has been approved, this PR is no longer useful.

**Pull request closed without merging**